Add Provisional GEP-1282 document (#1333)
kubernetes-sigs/gateway-api: site-src/geps/gep-1282.md

  * Linkerd uses a [Server resource](https://linkerd.io/2.11/reference/authorization-policy/#server) (which is functionally pretty similar to Service in that it associates a name with a Pod selector, but also has other details like if the service supports Proxy protocol), along with a ServerAuthorization resource that specifies some constructs that sit more at the service mesh level, including identity and access control.

In terms of other implementations existing use cases for features like this:
- For Contour, there are annotations to allow the configuration of the following on a backend Service:
  - Max connections
  - Max pending requests
  - Max requests
  - Max retries

>   - Upstream protocol: This is what allows Contour to handle the 

Reviewer: It looks like we didn't finish this list item?

- In Istio, the `DestinationRule` resource allows the configuration of many settings like this:
  - Load balancer algorithms
  - Connection pool settings
  - Outlier Detection
  - Tunnel settings for non-HTTP workloads

> 

Reviewer 2: ```suggestion
- In Consul, [`ServiceDefaults`](https://www.consul.io/docs/connect/config-entries/service-defaults) allows specifying similar configuration:
  - Max connections
  - Max pending requests
  - Max concurrent requests
  - Protocol
  - Passive health check interval and max failures (outlier detection)
  - TLS SNI
  
```

The properties we're talking about all share two things:
- They are tightly bound to the backend, rather than being important at the route level
- And the service owner (application developer in the Gateway API personas) should control these settings, not the owner of the Gateway implementation (the Cluster Operator)

> A good example is the load balancing algorithms for backends, as distinct from load balancing for routes. Load balancing for routes is about choosing which route option the implementation should choose, while load balancing for backends is about which endpoint from the backend the implementation should send traffic to if that backend's route is chosen. It's likely that these will need to be set differently (for example, random load balancing at the route level for canarying between services, with Weighted Least Request at the backend level for the actual endpoints).

Reviewer: This distinction is pretty important, and plays in with the mesh use case quite a bit. I tried to take a stab at this previously and we decided on policy attachment, it will be very important where we go from here. No action required, just highlighting this.

Reviewer 2: Worth cross-linking to prior discussion in https://github.com/kubernetes-sigs/gateway-api/pull/993 and https://github.com/kubernetes-sigs/gateway-api/issues/992?

Reviewer: sure

Author: I've added in references to this discussion.

In terms of other implementations existing use cases for features like this:
- For Contour, there are annotations to allow the configuration of the following on a backend Service:
  - Max connections
  - Max pending requests
  - Max requests
  - Max retries
  - Upstream protocol: This is what allows Contour to handle switching protocols for the backend service.
- In Istio, the `DestinationRule` resource allows the configuration of many settings like this:
  - Load balancer algorithms
  - Connection pool settings
  - Outlier Detection
  - Tunnel settings for non-HTTP workloads
- In Consul, [`ServiceDefaults`](https://www.consul.io/docs/connect/config-entries/service-defaults) allows specifying similar configuration:
  - Max connections
  - Max pending requests
  - Max concurrent requests
  - Protocol
  - Passive health check interval and max failures (outlier detection)
  - TLS SNI (we don't support configuring a specific CA at this granularity, but publishing a public key in a status field could be useful)
  
The properties we're talking about all share two things:

> - They are tightly bound to the backend, rather than being important at the route level

Reviewer 3: Maybe related to the caller as well

Author: They're relevant to the caller in how they take effect, but they are (or should be) _owned_ by the application owner, and apply to the service's pods and their associated endpoints, in Gateway API resource terms anyway.

Reviewer 3: Not exactly, in some case, A caller may use LB algorithm and connection pool settings different from caller B. How should gateway API allow this settings

> # GEP-696: GEP template

Reviewer: We would appear to want _some kind of title for this_?

Author: 🤦‍♂️

* Issue: [#1282](https://github.com/kubernetes-sigs/gateway-api/issues/1282)
* Status: Provisional

(See definitions in [Kubernetes KEP][kep-status].

[kep-status]: https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/kep.yaml#L9

## TLDR

> End-user feedback of Gateway API has presented a trend: some users are interested in defining functionality that describes capabilities of the service more than the route you take to get to the service. In order to support these more advanced route functions, implementations need programmatic access to deeper insights about a Service’s (that is: a Kubernetes Service that would be used in an *Route backendRef) capabilities, such as TLS or websocket details. The purpose of this GEP is to add APIs for capturing those Service capabilities to enable more complex routing composition.

Reviewer: > some users are interested in defining functionality that describes capabilities of the service more than the route you take to get to the service

To me this statement seems to demand more information about "why": it's not clear to me after reading the entire GEP so far _why_ users are interested in this functionality.

Author: It's a bit difficult to explain _why_ people want this things without talking in more detail about what they want, so I guess I'll need to fix this later in the GEP?

Bascially, the thing here is: Users have asked for stuff like backend TLS termination, and we don't have anywhere to put that in the current API. There are a few related features, so this GEP is about trying to make a broad solution to that class of problem - adding the description of extra capabilities to the backend.

I don't know how to say that more concisely than what I have here though.

Author: I've had a crack after updating the rest of the GEP, PTAL.

## Goals

* Provide a method to store properties and capabilities for service backends as extended-only features
* The method must be structured (that is, stored in a typed schema), and extensible (the design must include ways in which we can add more properties and capabilities in the future without redesigning it).
* The method should have a roadmap where we can take what we learn from this design and bring it to the core Kubernetes resources later.
* Clear separation of concerns with PolicyAttachment for properties which require multiple levels of defaults/overrides (timeouts, retries, etc).

## Non-Goals

* We don’t want to require changes to v1.Service to start with, as the turnaround time for those changes can be years.
* Our solution should not use unstructured data like labels and annotations - these are very sticky and hard to get rid of once you start using them.

> * v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the [IANA Service Name registry](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtmly), or domain-prefixed values and we need more flexibility than that.

Reviewer: good non-goals.

* To be populated as we discuss more.

## Introduction

As implementations have started to build out HTTPRoute support, requests for some common capabilities have started coming up, particularly TLS re-encryption (encryption between Gateway and backend), Websockets tracking, and HTTP/2 support.

Evan Anderson opened a [discussion that generated a lot of interest](https://github.com/kubernetes-sigs/gateway-api/discussions/1244), and during that discussion, we brought up a few more points:

> * Whatever we do to solve this in the HTTPRoute case has applicability to describing more general service properties, like identity, which could be useful for other Route use cases.

Reviewer: ```suggestion
* Whatever we do to solve this in the HTTPRoute case may be applicable to more general service properties, like identity, which could be useful for other Route use cases.
```

This initial form of this GEP is for the Gateway API community to agree on what problem we’re solving and why.

### What we're building

We’re looking to add specific, structured extension points somewhere in the resources used by the Gateway API to describe properties and capabilities of backend services.

The initial list includes, but is not limited to:
* TLS information for connection from the Gateway to the backend service. Note that this doesn’t include any information used for service mesh encryption, just what a Gateway’s proxy would need to be able to connect to the backend service.
* Websocket protocol information for the backend service.
* Protocol disambiguation for “upgradeable” protocols like HTTP/2 and HTTP/1.1 which operate on the same port. (Websockets may be another case of this.)

### Why build something?

The existing resources we have (Service and HTTPRoute) don’t have space for this information currently, and we have feedback from users via various implementations that it’s necessary - you could argue that some of this is table stakes for a layer 7 ingress controller.

> In previous discussions, we discussed a variety of ways to achieve this goal, but there’s been a lot of discussion already, so we’re moving to a GEP discussion format to try to capture feedback more efficiently.

Reviewer: Right now, I think this "why" section needs significantly more content, it doesn't seem to be fully supporting the "why should we do this" question:

> The existing resources we have (Service and HTTPRoute) don’t have space for this information currently

This just extends the "what": we're reiterating _what_ we want to see done by stating that it doesn't exist yet, but that isn't part of an answer to "why".

> and we have feedback from users via various implementations that it’s necessary

This comment needs some quantification, can we add the user feedback we've received here (even if paraphrased some) into some specific needs that justify the extra service metadata we're proposing?

> you could argue that some of this is table stakes for a layer 7 ingress controller.

This comment just seems to reiterate the desire of the author, but we need to quantify this: what capabilities of an L7 ingress controller demand the service metadata we're proposing?

>  In previous discussions, we discussed a variety of ways to achieve this goal, but there’s been a lot of discussion already, so we’re moving to a GEP discussion format to try to capture feedback more efficiently.

This doesn't seem to belong in "why": it seems to be more of a note to indicate there have been previous discussions (without links)?

Author: It seems like what we need here is:
- a list of the actual feature requests we have in this repo
- examples of how other ingress controllers have requests for similar things
- More links to evidence.

Does that sound right?

Reviewer: Sure, also potentially user stories. Ideally some stuff more written from the perspective of the cluster operator who has to set the ingress up for other teams.
